tokio/src/sync/once_cell.rs

```diff
@@ -399,6 +401,44 @@ impl<T> OnceCell<T> {
     pub fn take(&mut self) -> Option<T> {
         std::mem::take(self).into_inner()
     }
+
+    #[inline(always)]
+    /// Returns a future that resolves once a value has been written to the `OnceCell`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use tokio::sync::OnceCell;
+    ///
+    /// static LIFE: OnceCell<u8> = OnceCell::const_new();
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     tokio::task::spawn(async {
+    ///         let life: &u8 = LIFE.wait().await;
+    ///         assert_eq!(*life, 42);
+    ///     });
+    ///
+    ///     assert!(LIFE.set(42).is_ok());
+    /// }
+    /// ```
+    pub fn wait(&self) -> impl Future<Output = &T> {
+        self
+    }
+}
+
+// Future that resolves once a value has been written to the `OnceCell`.
+// Used in `OnceCell::wait`.
+impl<'a, T> Future for &'a OnceCell<T> {
+    type Output = &'a T;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.initialized() {
+            Poll::Ready(unsafe { self.get_unchecked() })
+        } else {
+            Poll::Pending
```

Comment:
If this doesn't immediately return ready, the task will never wake, as the waker is lost.